File: leetcode/Binary Tree Preorder Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        order = []
        if root:
            stack = [root]
            while stack:
                node = stack.pop()
                order.append(node.val)
                if node.right:
                    stack.append(node.right)
                if node.left:
                    stack.append(node.left)
        return order
            
            
            

File: leetcode/cpp/Binary_Tree_Preorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        //非递归解法
        vector<int> res;
        stack<TreeNode*> stk;
        if (root)
            stk.push(root);
        while (!stk.empty()) {
            root = stk.top();
            stk.pop();
            res.push_back(root->val);
            if (root->right)
                stk.push(root->right);
            if (root->left)
                stk.push(root->left);
        }
        return res;
    }
};